### PR TITLE
:sparkles: feat: Add gas estimation actions

### PR DESCRIPTION
**Pull request by roninjin10**
- Copy paste gas estimation actions from https://github.com/ethereum-optimism/optimism/blob/develop/packages/fee-estimation/src/estimateFees.spec.ts
- Changed naming conventions to match op-viem
- deleted some of the more niche actions like `version` and `l1BaseFee`. The recommended way to use this is to just use the gasPriceOracle contract or get stuff like l1BaseFee from l1 they just bloat the API including them here imo

### CODE REVIEW

File: src/decorators/publicL2OpStackActions.ts [hunk 1]
```diff
@@ -20,6 +22,23 @@ export type PublicL2OpStackActions<
...
     args: getProveWithdrawalTransactionArgsParams,
   ) => Promise<getProveWithdrawalTransactionArgsReturnType>
+  /**
+   * Estimate the l1 gas price portion for a transaction
+   * @example
+   * const price = await getL1GasPrice(publicClient, {
+   *   blockNumber,
+   *   blockTag,
+   * });
+   */
```
Comment (roninjin10) [inline src/decorators/publicL2OpStackActions.ts:32]:
```suggestion
```
Maybe delete for now and go through and add jsdocs in it's own pr

File: package.json [hunk 2]
```diff
@@ -13,6 +13,7 @@
     "@viem/anvil": "^0.0.6",
     "@vitest/coverage-v8": "^0.34.2",
     "@wagmi/cli": "^1.3.0",
+    "abitype": "^0.9.8",
```
Comment (wilsoncusack) [inline package.json:16]: do we need this or do we get it already from viem?

File: src/actions/public/L2/feeActions.spec.ts
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.spec.ts:1]: we used `.test`

File: src/actions/public/L2/feeActions.spec.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    createPublicClient: (...args: [any]) => {
+      const client = _viem.createPublicClient(...args)
+      client.getGasPrice = async () => parseGwei('0.00000042')
+      return client
+    },
+  }
+})
+
+// using this optimist https://optimistic.etherscan.io/tx/0xaa291efba7ea40b0742e5ff84a1e7831a2eb6c2fc35001fa03ba80fd3b609dc9
+const blockNumber = BigInt(107028270)
```
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.spec.ts:25]:
```suggestion
const blockNumber = 107028270n
```

File: src/actions/public/L2/feeActions.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  TransactionSerializableEIP1559,
+  TransactionSerializedEIP1559,
+  Transport,
+} from 'viem'
+import { type Chain } from 'viem/chains'
+
+/**
+ * Bytes type representing a hex string with a 0x prefix
+ */
+type Bytes = `0x${string}`
```
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.ts:21]: we already have Hex and Bytes from viem?

File: src/actions/public/L2/feeActions.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  TFunctionName extends string | undefined = undefined,
+>(
+  client: PublicClient<Transport, TChain>,
+  options: OracleTransactionParameters<TAbi, TFunctionName>,
+) => Promise<bigint>
+
+/**
+ * Internal helper to serialize a transaction
+ */
+const transactionSerializer = <
```
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.ts:67]: would prefer we keep to one function per file, helps keep testing clear and some of these things may be handy elsewhere!

File: src/actions/public/L2/feeActions.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+ * @example
+ * const contract = getGasPriceOracleContract(publicClient)
+ * const gasPrice = await contract.read.gasPrice({ blockNumber })
+ */
+const getGasPriceOracleContract = <TChain extends Chain | undefined>(
+  client: PublicClient<Transport, TChain>,
+) => {
+  return getContract({
+    // it's a predeploy so it's the same on all chains
+    address: gasPriceOracleAddress['420'],
```
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.ts:95]:
```suggestion
    address: opStackL2ChainContracts.gasPriceOracle,
```

File: src/actions/public/L2/feeActions.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  return serializedTransaction as TransactionSerializedEIP1559
+}
+
+/**
+ * Get gas price Oracle contract
+ * @example
+ * const contract = getGasPriceOracleContract(publicClient)
+ * const gasPrice = await contract.read.gasPrice({ blockNumber })
+ */
+const getGasPriceOracleContract = <TChain extends Chain | undefined>(
```
Comment (wilsoncusack) [inline src/actions/public/L2/feeActions.ts:90]: again prefer 1 function per file. 

Thinking if we should actually export these viem versions of all the main contracts 🤔

File: src/decorators/publicL2OpStackActions.ts (hunk 1 above)
Comment (wilsoncusack) [reply src/decorators/publicL2OpStackActions.ts:32]: yeah I've been underinvesting in jsdoc in favor of the markdown docs, not sure if right tradeoff.

File: package.json (hunk 2 above)
Comment (roninjin10) [reply package.json:16]: I think can get it from viem

File: src/actions/public/L2/feeActions.spec.ts [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+import { estimateFees, estimateL1Fee, estimateL1GasUsed } from './feeActions'
+
+vi.mock('viem', async () => {
+  const _viem = (await vi.importActual('viem')) as any
+  return {
+    ..._viem,
+    // no way to get historical gas price
+    createPublicClient: (...args: [any]) => {
+      const client = _viem.createPublicClient(...args)
+      client.getGasPrice = async () => parseGwei('0.00000042')
```
[the remark this answers is not in the document]
Comment (roninjin10) [reply src/actions/public/L2/feeActions.spec.ts:18]: It does not unfortunately. Viem types correctly don't give you blockTag as a param as well

File: src/actions/public/L2/feeActions.spec.ts (hunk 3 above)
Comment (roninjin10) [reply src/actions/public/L2/feeActions.spec.ts:25]: Keeping open because if I merge it into github remote I might accidentally force push the change away but will do

File: src/actions/public/L2/feeActions.ts (hunk 7 above)
Comment (roninjin10) [reply src/actions/public/L2/feeActions.ts:90]: @evmts/contracts-ts is the package that exports versions of all the main contracts. It does it both as react hooks and as vanilla hooks. It might be a good idea to reexport them from this package though

File: src/actions/public/L2/feeActions.ts (hunk 6 above)
Comment (roninjin10) [reply src/actions/public/L2/feeActions.ts:95]: Will do keeping comment open so i don't accidentally force push away

File: src/decorators/publicL2OpStackActions.ts (hunk 1 above)
Comment (roninjin10) [reply src/decorators/publicL2OpStackActions.ts:32]: Jsdoc is more important because you see jsdoc in your editor imo

File: src/actions/public/L2/feeActions.spec.ts [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+const paramsWithdraw = {
+  blockNumber: BigInt(107046472),
+} as const
+
+beforeEach(() => {
+  vi.resetAllMocks()
+})
+
+test('estimateFees should return correct fees', async () => {
+  // burn
```
Comment (roninjin10) [inline src/actions/public/L2/feeActions.spec.ts:82]: remove the copilot comments in the test

File: src/actions/public/L2/feeActions.spec.ts [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+const viemClient = createPublicClient({
+  chain: optimism,
+  transport: http(process.env.VITE_L2_RPC_URL ?? 'https://mainnet.optimism.io'),
+})
+
+const paramsOptimist = {
+  blockNumber,
+} as const
+const paramsWithdraw = {
+  blockNumber: BigInt(107046472),
```
Comment (roninjin10) [inline src/actions/public/L2/feeActions.spec.ts:74]: 134134n instead of BigInt

File: src/decorators/publicL2OpStackActions.ts [hunk 11]
```diff
@@ -31,5 +50,8 @@ export function publicL2OpStackActions<
   return {
     getWithdrawalMessages: (args) => getWithdrawalMessages(client, args),
     getProveWithdrawalTransactionArgs: (args) => getProveWithdrawalTransactionArgs(client, args),
+    estimateL1Fee: (args) => estimateL1Fee(client, args),
+    estimateL1GasUsed: (args) => estimateL1GasUsed(client, args),
+    estimateFees: (args) => estimateFees(client, args),
```
Comment (roninjin10) [inline src/decorators/publicL2OpStackActions.ts:55]: @wilsoncusack what do you think about making this look like this?
```typescript
export function publicL2OpStackActions = createDecorator([getWithdrawalMessages, getProveWithdrawalTransactionArgs, estimateL1Fee...])
```

Comment (roninjin10) [reply src/decorators/publicL2OpStackActions.ts:55]: bad idea lose jsdoc if we do this

File: src/actions/public/L2/feeActions.spec.ts (hunk 10 above)
Comment (roninjin10) [reply src/actions/public/L2/feeActions.spec.ts:74]:
```suggestion
  blockNumber: 107046472n,
```

File: src/actions/public/L2/estimateFees.test.ts [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    // extra data
+    '0x0',
+  ],
+  maxFeePerGas: parseGwei('.2'),
+  maxPriorityFeePerGas: parseGwei('.1'),
+} as const
+
+const viemClient = createPublicClient({
+  chain: optimism,
+  transport: http(process.env.VITE_L2_RPC_URL ?? 'https://mainnet.optimism.io'),
```
Comment (roninjin10) [inline src/actions/public/L2/estimateFees.test.ts:68]:
```suggestion
  transport: http(process.env.VITE_OP_MAINNET_ROLLUP_FORK_URL ?? 'https://mainnet.optimism.io'),
```

File: src/actions/public/L2/estimateFees.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ * on non OP chains this is usually GasUsed * GasPrice
+ * on OP chains this is GasUsed * GasPrice + L1Fee
+ * @example
+ * const feeValue = await estimateFees(publicClient, {
+ *   abi,
+ *   functionName: balanceOf,
+ *   args: [address],
+ * });
+ */
+export const estimateFees: EstimateFees = async (client, options) => {
```
Comment (wilsoncusack) [inline src/actions/public/L2/estimateFees.ts:41]: in an ideal world, this would be the default `estimateGas` function for an OPStack chain in viem, right? @roninjin10 

Do we think we should name it like opStackEstimateGas or something closer?

File: src/types/gasPriceOracle.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  PublicClient,
+  TransactionSerializableEIP1559,
+  Transport,
+} from 'viem'
+import type { Chain } from 'viem/chains'
+
+/**
+ * Options to query a specific block
+ */
+type BlockOptions = {
```
Comment (wilsoncusack) [inline src/types/gasPriceOracle.ts:14]: why not export this and use it in the other files instead of re stating it? 🤔

File: src/actions/public/L2/estimateFees.test.ts [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (wilsoncusack) [inline src/actions/public/L2/estimateFees.test.ts:68]: will this only work with optimism? If so, I think you don't want to allow VITE_L2_RPC_URL, if it'll work with others, you should use https://github.com/base-org/op-viem/blob/main/src/actions/public/L2/getWithdrawalMessages.test.ts#L2

File: src/actions/public/L2/estimateFees.ts (hunk 13 above)
Comment (roninjin10) [reply src/actions/public/L2/estimateFees.ts:41]: Estimate gas is different. On l1 estimate fees would be estimateGas multiplied by the gas price. This is something I chatted with Jake about and he lightly agreed it would be valuable to add for all chains

File: src/types/gasPriceOracle.ts (hunk 14 above)
Comment (roninjin10) [reply src/types/gasPriceOracle.ts:14]: Can do this we just landed here from a refactor to make it one function per file

File: src/actions/public/L2/estimateFees.test.ts (hunk 15 above)
Comment (roninjin10) [reply src/actions/public/L2/estimateFees.test.ts:68]: Should append the chain Id to the env variable name imo it only works with op

File: src/actions/public/L2/estimateFees.ts (hunk 13 above)
Comment (wilsoncusack) [reply src/actions/public/L2/estimateFees.ts:41]: ok thanks for clarifying @roninjin10

File: src/types/gasPriceOracle.ts (hunk 14 above)
Comment (wilsoncusack) [reply src/types/gasPriceOracle.ts:14]: Roger yeah appreciate you breaking out, just would prefer to DRY up types if we can.

File: src/actions/public/L2/estimateFees.test.ts (hunk 15 above)
Comment (wilsoncusack) [reply src/actions/public/L2/estimateFees.test.ts:68]: > Should append the chain Id to the env variable name imo

sorry can you say what you mean? Like VITE_8453_RPC_URL?

File: src/actions/public/L2/estimateFees.ts (hunk 13 above)
Comment (roninjin10) [reply src/actions/public/L2/estimateFees.ts:41]: I’m not in love with the name estimateFees btw I just couldn’t come up with a better name.  Etherscan calls it a transactionFee. Maybe we should call it transactionFee? I would love a name that makes it so folks can guess the difference between it and estimateGas most of the time

File: src/actions/public/L2/estimateFees.test.ts (hunk 15 above)
Comment (wilsoncusack) [reply src/actions/public/L2/estimateFees.test.ts:68]: hmm I think would break precedent with viem? sees ideally there's one rollup fork and one L1 fork?

Comment (roninjin10) [reply src/actions/public/L2/estimateFees.test.ts:68]: Yea we can keep iterating on it. Just for historical context this test isn't meant to be what gets upstreamed from viem it's just the test that happened to get written when this originally landed in the optimism monorepo. We should definitely replace it with a higher quality test.

A project we are kicking off now is the ability to test something like this using a devnet like anvil which this (and all of op-viem) would be a good use case for dogfooding.  One of the requirements of this project is to be able to test things like l1 gas fee against a devnet

File: src/actions/public/L2/estimateL1GasUsed.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+ * const L1GasUsedValue = await estimateL1GasUsed(data, {
+ *  abi,
+ *  functionName: balanceOf,
+ *  args: [address],
+ * });
+ */
+export const estimateL1GasUsed: GasPriceOracleEstimator = async (client, options) => {
+  const data = serializeEip1559Transaction(options)
+  return readContract(client, {
+    address: gasPriceOracleAddress['420'],
```
[the remark this answers is not in the document]
Comment (roninjin10) [reply src/actions/public/L2/estimateL1GasUsed.ts:67]: I would prefer we set it up to always pull the contract addresses from the monorepo. Both l1 and l2 but more importantly l1 since l2 are just predeployed and the same on all op chains

[the remark this answers is not in the document]
Comment (wilsoncusack) [reply src/actions/public/L2/estimateL1GasUsed.ts:67]: ok but @roninjin10 this currently breaks precedent with the rest of the repo ([e.g.](https://github.com/base-org/op-viem/blob/main/src/actions/wallet/L2/writeWithdrawERC20.ts#L31)), so let's follow precedent for now and switch them all if we decided to.

File: src/actions/public/L2/estimateFees.ts (hunk 13 above)
Comment (wilsoncusack) [reply src/actions/public/L2/estimateFees.ts:41]: transactionFee is pretty good! I don't have a strong preference

**Review by wilsoncusack — COMMENTED** [review]
@roninjin10 we should also export this in actions/index

File: src/actions/index.ts [hunk 17]
```diff
@@ -36,6 +36,9 @@ export {
   type SimulateProveWithdrawalTransactionParameters,
   type SimulateProveWithdrawalTransactionReturnType,
 } from './public/L1/simulateProveWithdrawalTransaction.js'
+export { estimateFees, type EstimateFeesParameters } from './public/L2/estimateFees.js'
+export { estimateL1Fee, type EstimateL1FeeParameters } from './public/L2/estimateL1Fee.js'
+export { estimateL1GasUsed, type GasPriceOracleEstimator } from './public/L2/estimateL1GasUsed.js'
```
Comment (roninjin10) [inline src/actions/index.ts:41]: This looks like a bug
```suggestion
export { estimateL1GasUsed, type EstimateL1GasUsedParameters } from './public/L2/estimateL1GasUsed.js'
```

**Review by zencephalon — APPROVED** [review]
Approving with fast-follows for documentation and testing updates.
